sympy/core/basic.py

```diff
@@ -1623,32 +1623,72 @@ def count_ops(self, visual=None):
         from sympy import count_ops
         return count_ops(self, visual)
 
-    def doit(self, **hints):
-        """Evaluate objects that are not evaluated by default like limits,
-        integrals, sums and products. All objects of this kind will be
-        evaluated recursively, unless some species were excluded via 'hints'
-        or unless the 'deep' hint was set to 'False'.
+    def doit(self, deep=True, properties=[], **hints):
+        """Evaluate objects that are not evaluated like limits, integrals,
+        sums and products.
 
-        >>> from sympy import Integral
-        >>> from sympy.abc import x
+        Arguments will be evaluated recursively, unless some classes were
+        excluded via 'properties' or unless the 'deep' hint was set to 'False'.
+
+        Evaluatable classes should define their own `_eval_doit` method,
+        which returns evaluated result. In this case, non-recursive evaluation
+        should also be done when `deep=False` is passed. It is also encouraged
+        to allow `evaluate` option to the constructor.
 
-        >>> 2*Integral(x, x)
-        2*Integral(x, x)
+        >>> from sympy import Add, Integral, Derivative
+        >>> from sympy.abc import x
 
-        >>> (2*Integral(x, x)).doit()
-        x**2
+        >>> expr = Add(Integral(x, x), Integral(x, x), Derivative(x, x), evaluate=False)
+        >>> expr
+        Derivative(x, x) + Integral(x, x) + Integral(x, x)
 
-        >>> (2*Integral(x, x)).doit(deep=False)
-        2*Integral(x, x)
+        >>> expr.doit()
+        x**2 + 1
 
+        >>> expr.doit(deep=False)
+        Derivative(x, x) + 2*Integral(x, x)
         """
-        if hints.get('deep', True):
-            terms = [term.doit(**hints) if isinstance(term, Basic) else term
-                                         for term in self.args]
+        hints.update(deep=deep, properties=properties)
+
+        if deep:
+            terms = [term.doit(**hints) if isinstance(term, Basic)
+                        else term for term in self.args]
+        else:
+            terms = self.args
+
+        eval_doit = self._eval_doit(*terms, **hints)
+        if eval_doit is not None:
+            return eval_doit
+
+        # Place this condition here so that Atom's subclass
+        # can define its own _eval_doit.
+        if self.is_Atom:
+            return self
+
+        if deep:
             return self.func(*terms)
         else:
             return self
 
+    def _eval_doit(self, *args, **hints):
+        """
+        If `deep=True` is passed to `doit` which called this method,
```

Comment:
If this method receives the args as an argument then I guess it can be a class method and then it can be called from `__new__`

Reply:
Whether ot not `_eval_doit` is a class method is a key point that needs to be resolved.

Reply:
Agree. I am also thinking to use the name `eval` (as `Function` does) instead of `_eval_doit`